util/rwlock/src/rwlock.c

```diff
@@ -0,0 +1,163 @@
+
+/**
+ * Unblocks the next pending user.  The caller must lock the mutex prior to
+ * calling this.
+ */
+static void
+rwlock_unblock(struct rwlock *lock)
+{
+    /* Give priority to pending writers. */
+    if (lock->pending_writers > 0) {
+        os_sem_release(&lock->wsem);
+        lock->pending_writers--;
+    } else {
+        while (lock->pending_readers > 0) {
+            os_sem_release(&lock->rsem);
+            lock->pending_readers--;
+        }
+    }
+}
+
+int
+rwlock_acquire_read(struct rwlock *lock, os_time_t timeout)
+{
+    bool acquired;
+    int rc;
+
+    os_mutex_pend(&lock->mtx, OS_TIMEOUT_NEVER);
+
+    /* Readers have low priority; if there is an active writer or a pending
+     * writer, we have to block.
+     */
+    if (lock->active_writer || lock->pending_writers > 0) {
+        lock->pending_readers++;
+        acquired = false;
+    } else {
+        lock->num_readers++;
+        acquired = true;
+    }
+
+    os_mutex_release(&lock->mtx);
+
+    if (acquired) {
+        /* No contention; lock acquired. */
+        return 0;
+    }
+
+    /* Wait for the lock to become available. */
+    rc = os_sem_pend(&lock->rsem, timeout);
+
+    os_mutex_pend(&lock->mtx, OS_TIMEOUT_NEVER);
+    if (rc != 0 && rc != OS_NOT_STARTED) {
+        /* Semaphore timed out; undo increment and return error. */
+        lock->pending_readers--;
+        rc = SYS_ETIMEOUT;
+    } else {
+        /* Success; record reader ownership. */
+        lock->num_readers++;
+        rc = 0;
+    }
+    os_mutex_release(&lock->mtx);
+
+    return rc;
+}
+
+void
+rwlock_release_read(struct rwlock *lock)
+{
+    os_mutex_pend(&lock->mtx, OS_TIMEOUT_NEVER);
+
+    assert(lock->num_readers > 0);
+    lock->num_readers--;
+
+    /* If this is the last active reader, unblock a pending writer if there is
+     * one.
+     */
+    if (lock->num_readers == 0) {
+        rwlock_unblock(lock);
+    }
+
+    os_mutex_release(&lock->mtx);
+}
+
+int
+rwlock_acquire_write(struct rwlock *lock, os_time_t timeout)
+{
+    bool acquired;
+    int rc;
+
+    os_mutex_pend(&lock->mtx, OS_TIMEOUT_NEVER);
+
+    /* If there are any active readers or an active writer, we have to wait. */
+    if (lock->num_readers > 0 || lock->active_writer) {
+        lock->pending_writers++;
+        acquired = false;
+    } else {
+        lock->active_writer = true;
+        acquired = true;
+    }
+
+    os_mutex_release(&lock->mtx);
+
+    if (acquired) {
+        /* No contention; lock acquired. */
+        return 0;
+    }
+
+    /* Wait for the lock to become available. */
+    rc = os_sem_pend(&lock->wsem, timeout);
+
+    os_mutex_pend(&lock->mtx, OS_TIMEOUT_NEVER);
+    if (rc != 0 && rc != OS_NOT_STARTED) {
+        /* Semaphore timed out; undo increment and return error. */
+        lock->pending_writers--;
+        rc = SYS_ETIMEOUT;
+    } else {
+        /* Success; record writer ownership. */
+        lock->active_writer = true;
+        rc = 0;
```

Comment:
This also should re-evaluate lock state once sem_pend() has returned, after lock->mtx was acquired.